fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/DefaultGranularityGenerator.java

```diff
@@ -0,0 +1,134 @@
+// Copyright 2019 Oath Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.web.apirequest.generator;
+
+import static com.yahoo.bard.webservice.web.ErrorMessageFormat.UNKNOWN_GRANULARITY;
+import static com.yahoo.bard.webservice.web.apirequest.DataApiRequestBuilder.RequestResource.GRANULARITY;
+import static com.yahoo.bard.webservice.web.apirequest.DataApiRequestBuilder.RequestResource.TIMEZONE;
+
+import com.yahoo.bard.webservice.data.time.Granularity;
+import com.yahoo.bard.webservice.data.time.GranularityParser;
+import com.yahoo.bard.webservice.util.GranularityParseException;
+import com.yahoo.bard.webservice.web.BadApiRequestException;
+import com.yahoo.bard.webservice.web.apirequest.DataApiRequest;
+import com.yahoo.bard.webservice.web.apirequest.DataApiRequestBuilder;
+import com.yahoo.bard.webservice.web.apirequest.RequestParameters;
+import com.yahoo.bard.webservice.web.util.BardConfigResources;
+
+import org.joda.time.DateTimeZone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * Default generator implementation for binding {@link Granularity}. Granularity is dependent on {@link DateTimeZone}
+ * already being bound. Ensure the generator for DateTimeZone is called before attempting to bind granularity with this
+ * generator.
+ */
+public class DefaultGranularityGenerator implements Generator<Granularity> {
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultGranularityGenerator.class);
+
+    /**
+     * Binds request granularity into a {@link Granularity} object.
+     *
+     * Throws BadApiRequestException if granularity is not present. Granularity is required in all data queries. Also
+     * is thrown if an invalid granularity is provided.
+     * Throws UnsatisfiedApiRequestConstraintsException if timeZone was not built before this generator runs. Timezone
+     * CAN be empty, which implies using the default timezone. A NULL timezone indicates timezone has not been
+     * generated yet.
+     *
+     * @param builder  The builder object representing the in progress {@link DataApiRequest}. Previously constructed
+     *        resources are available through this object.
+     * @param params  The request parameters sent by the client.
+     * @param resources  Resources used to build the request, such as the
+     *        {@link com.yahoo.bard.webservice.data.config.ResourceDictionaries}.
+     * @return the generated granularity.
+     */
+    @Override
+    public Granularity bind(
+            DataApiRequestBuilder builder,
+            RequestParameters params,
+            BardConfigResources resources
+    ) {
+        String granularity = params.getGranularity()
+                .orElseThrow(() -> new BadApiRequestException(UNKNOWN_GRANULARITY.logFormat("null")));
```

Comment:
I'm fairly convinced that this 'null' isn't the right messaging here.

I'm trying to trace the code, is there a code path going through here?

Reply:
I will change the message to something else. I was mostly trying to duplicate https://github.com/yahoo/fili/blob/master/fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ApiRequestImpl.java#L237-L248 , but after looking at it again I think I just need to say granularity was missing from the query.

Reply:
ultimately, we're going to probably treat missing granularity as meaning all granularity eventually.

Is there any code path that does down this path?

Reply:
Currently all data requests must have granularity to even hit the data servlet. Under normal circumstances this code path should never be hit.